File: app/src/main/java/de/ct/nutria/foodSelector/FoodItemFragment.java
package de.ct.nutria.foodSelector;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.SimpleAdapter;
import android.widget.TextView;

import de.ct.nutria.R;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * A fragment displaying a list of FoodItems.
 * <p/>
 * Activities containing this fragment MUST implement the {@link OnListSelect}
 * interface.
 */
public class FoodItemFragment extends Fragment implements AdapterView.OnItemClickListener {
    private String PARCELABLE_FOOD_LIST = "de.ct.nutria.foodSelector.FoodItemFragment.foodList";
    private OnListSelect listSelectListener;
    private ArrayList<FoodItem> foodList;
    private ListView listView;
    private TextView emptyView;
    private EditText searchEntry;
    private ImageButton searchButton;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public FoodItemFragment() {
    }

    public static FoodItemFragment newInstance() {
        FoodItemFragment fragment = new FoodItemFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        foodList = new ArrayList<>();
        if (getArguments() != null) {
            // Read arguments from Bundle
            foodList = savedInstanceState.getParcelableArrayList(PARCELABLE_FOOD_LIST);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_fooditem_list, container, false);
        listView = view.findViewById(R.id.foodList);
        listView.setOnItemClickListener(this);
        emptyView = view.findViewById(R.id.foodListEmpty);
        searchEntry = view.findViewById(R.id.searchEntry);
        searchEntry.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                requestList(s);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        searchButton = view.findViewById(R.id.searchButton);
        return view;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListSelect) {
            listSelectListener = (OnListSelect) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listSelectListener = null;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);
        savedInstanceState.putParcelableArrayList(PARCELABLE_FOOD_LIST, foodList);
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        Log.d("Item clicked with no", Integer.toString(i));
        this.listSelectListener.onListSelect(foodList.get(i));
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    public interface OnListSelect {
        void onListSelect(FoodItem item);
    }

    private void requestList(CharSequence query) {
        FoodItem test = new FoodItem("Foo");
        test.setType(1);
        test.setAuthorName("Maria Musterfrau");
        test.setCategory(1, "test");
        test.setCalories(423);
        Log.d(test.getName(), test.getCaloriesString());
        Log.d("Manufacturer", test.describeManufacturer("a ", "b", "c"));
        for(int i=0; i < foodList.size(); i++) {
            if(!foodList.get(i).getName().contains(query)) {
                foodList.remove(i);
                i--;
            }
        }
        requestQueryFromNutriaDb(query);
        updateList();
    }

    private void updateList() {
        ArrayList<HashMap<String,String>> displayFoodList=new ArrayList<>();
        for (FoodItem item : foodList) {
            HashMap<String, String> entry = new HashMap<>();
            entry.put("foodName", item.getName());
            entry.put("manufacturer", item.describeManufacturer(
                    getString(R.string.recipe_by),
                    getString(R.string.selfmade),
                    getString(R.string.harvested)));
            entry.put("calories", item.getCaloriesString());
            entry.put("iconID", getIcon(item.getCategoryId())+"");
            displayFoodList.add(entry);
        }
        String[] fromArray = {"foodName", "manufacturer",
                              "calories", "iconID"};
        int[] viewIndexes = {R.id.listFoodName, R.id.listManufacturer,
                             R.id.listCalories, R.id.listCategoryIcon};
        SimpleAdapter adapter = new SimpleAdapter(getActivity(), displayFoodList,
                R.layout.fragment_fooditem, fromArray, viewIndexes);
        listView.setAdapter(adapter);
        listView.setVisibility((displayFoodList.size() > 0) ? View.VISIBLE : View.INVISIBLE);
        emptyView.setVisibility((displayFoodList.size() <= 0) ? View.VISIBLE : View.INVISIBLE);
    }

    private int getIcon(int categoryId) {
        if (categoryId == 1) return R.drawable.ic_category_1;
        if (categoryId == 2) return R.drawable.ic_category_2;
        if (categoryId == 3) return R.drawable.ic_category_3;
        if (categoryId == 6) return R.drawable.ic_category_6;
        return R.drawable.ic_tupperdose;
    }

    private void requestQueryFromNutriaDb(CharSequence query) {
        createDummyList();
    }

    private void createDummyList() {
        foodList.clear();
        FoodItem i1 = new FoodItem("Banane");
        i1.setType(0);
        i1.setCategory(1, "Obst");
        i1.setCalories(123);
        i1.setAuthorName("Max Mustermann");
        foodList.add(i1);
        FoodItem i2 = new FoodItem("Apfel");
        i2.setType(0);
        i2.setCategory(6, "Obst");
        i2.setCalories(82);
        i2.setAuthorName("Max Mustermann");
        foodList.add(i2);
        FoodItem i3 = new FoodItem("Ritter Sport - Honig-Salz-Mandel");
        i3.setType(0);
        i3.setManufacturer("Ritter Sport");
        i3.setCategory(2, "Schokolade");
        i3.setCalories(564);
        i3.setAuthorName("Andreas Adipositas");
        foodList.add(i3);
        FoodItem i4 = new FoodItem("Salz");
        i4.setType(0);
        i4.setManufacturer("Bad Reichenhaller");
        i4.setCategory(3, "Gewürz");
        i4.setCalories(1);
        i4.setAuthorName("Andreas Adipositas");
        foodList.add(i4);
    }
}

File: app/src/main/kotlin/foodSelector/FoodItemFragment.kt
package foodSelector

import android.app.Fragment
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.EditText
import android.widget.ListView
import android.widget.SimpleAdapter

import de.ct.nutria.R
import de.ct.nutria.foodSelector.NutriaRequestCallback
import kotlinx.android.synthetic.main.fragment_fooditem_list.*

import java.util.ArrayList
import java.util.HashMap

/**
 * A fragment displaying a list of FoodItems.
 *
 *
 * Activities containing this fragment MUST implement the [OnListSelect]
 * interface.
 */
/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */
class FoodItemFragment : Fragment(), AdapterView.OnItemClickListener, NutriaRequestCallback<String> {
    private val PARCELABLE_FOOD_LIST = "foodSelector.FoodItemFragment.foodArray"
    private var listSelectListener: OnListSelect? = null
    private var foodArray: ArrayList<FoodItem>? = null
    //private var nutriaNetworkFragment: NutriaHeadlessNetworkFragment? = null
    private var nutriaRequestOngoing = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        foodArray = ArrayList()
        if (arguments != null && savedInstanceState != null) {
            // Read arguments from Bundle
            foodArray = savedInstanceState!!.getParcelableArrayList(PARCELABLE_FOOD_LIST)
        }
        //nutriaNetworkFragment = NutriaHeadlessNetworkFragment.getInstance(
        //        activity.fragmentManager)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_fooditem_list, container, false)
        view.findViewById<ListView>(R.id.foodList)?.onItemClickListener = this
        view.findViewById<EditText>(R.id.searchEntry)?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                requestList(s)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
        return view
    }


    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnListSelect) {
            listSelectListener = context
        } else {
            throw RuntimeException("$context must implement OnListFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listSelectListener = null
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        savedInstanceState.putParcelableArrayList(PARCELABLE_FOOD_LIST, foodArray)
    }

    override fun onItemClick(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        Log.d("Item clicked with no", Integer.toString(i))
        if (this.listSelectListener != null && foodArray != null)
            this.listSelectListener!!.onListSelect(foodArray!![i])
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    interface OnListSelect {
        fun onListSelect(item: FoodItem)
    }

    private fun requestList(query: CharSequence) {
        val test = FoodItem("Foo")
        test.type = 1
        test.authorName = "Maria Musterfrau"
        test.setCategory(1, "test")
        test.calories = 423.0f
        Log.d(test.name, test.caloriesString)
        Log.d("Manufacturer", test.describeManufacturer("a ", "b", "c"))
        var i = 0
        if (foodArray != null) {
            while (i < foodArray!!.size) {
                if (!foodArray!![i].name.contains(query)) {
                    foodArray!!.removeAt(i)
                    i--
                }
                i++
            }
        }
        requestQueryFromNutriaDb(query)
        updateList()
    }

    private fun updateList() {
        val displayFoodList = ArrayList<HashMap<String, String>>()
        if (foodArray != null) for (item in foodArray!!) {
            val entry = HashMap<String, String>()
            entry["foodName"] = item.name
            entry["manufacturer"] = item.describeManufacturer(
                    getString(R.string.recipe_by),
                    getString(R.string.selfmade),
                    getString(R.string.harvested))
            entry["calories"] = item.caloriesString
            entry["iconID"] = getIcon(item.categoryId).toString() + ""
            displayFoodList.add(entry)
        }
        val fromArray = arrayOf("foodName", "manufacturer", "calories", "iconID")
        val viewIndexes = intArrayOf(R.id.listFoodName, R.id.listManufacturer, R.id.listCalories, R.id.listCategoryIcon)
        val adapter = SimpleAdapter(activity, displayFoodList,
                R.layout.fragment_fooditem, fromArray, viewIndexes)
        foodList.adapter = adapter
        foodList.visibility = if (displayFoodList.size > 0) View.VISIBLE else View.INVISIBLE
        foodListEmpty.visibility = if (displayFoodList.size <= 0) View.VISIBLE else View.INVISIBLE
    }

    private fun getIcon(categoryId: Int): Int {
        if (categoryId == 1) return R.drawable.ic_category_1
        if (categoryId == 2) return R.drawable.ic_category_2
        if (categoryId == 3) return R.drawable.ic_category_3
        return if (categoryId == 6) R.drawable.ic_category_6 else R.drawable.ic_tupperdose
    }

    private fun requestQueryFromNutriaDb(query: CharSequence) {
        /*if (!this.nutriaRequestOngoing && this.nutriaNetworkFragment != null) {
            try {
                this.nutriaNetworkFragment.doQuery(query);
                this.nutriaRequestOngoing = true;
            } catch (JSONException e) {
                Log.e("JSON error in payload", e.getMessage());
            }
        }*/ //TODO: Decide what to do if a request is already running.
        createDummyList()
    }

    override fun updateFromRequest(result: String) {

    }

    override fun getActiveNetworkInfo(): NetworkInfo {
        val connectivityManager = activity.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        return connectivityManager.activeNetworkInfo
    }

    override fun onProgressUpdate(progressCode: Int) {
        when (progressCode) {
            // You can add UI behavior for progress updates here.
            NutriaRequestCallback.Progress.ERROR -> {
            }
            NutriaRequestCallback.Progress.CONNECT_SUCCESS -> {
            }
            NutriaRequestCallback.Progress.GET_INPUT_STREAM_SUCCESS -> {
            }
            NutriaRequestCallback.Progress.PROCESS_INPUT_STREAM_IN_PROGRESS -> {
            }
            NutriaRequestCallback.Progress.PROCESS_INPUT_STREAM_SUCCESS -> {
            }
        }
    }

    override fun finishRequest() {
        this.nutriaRequestOngoing = false
        /*if (this.nutriaNetworkFragment != null) {
            this.nutriaNetworkFragment!!.cancelRequest()
        }*/
    }

    private fun createDummyList() {
        if (foodArray == null) foodArray = ArrayList()
        foodArray!!.clear()
        val i1 = FoodItem("Banane")
        i1.type = 0
        i1.setCategory(1, "Obst")
        i1.calories = 123f
        i1.authorName = "Max Mustermann"
        foodArray!!.add(i1)
        val i2 = FoodItem("Apfel")
        i2.type = 0
        i2.setCategory(6, "Obst")
        i2.calories = 82f
        i2.authorName = "Max Mustermann"
        foodArray!!.add(i2)
        val i3 = FoodItem("Ritter Sport - Honig-Salz-Mandel")
        i3.type = 0
        i3.manufacturer = "Ritter Sport"
        i3.setCategory(2, "Schokolade")
        i3.calories = 564f
        i3.authorName = "Andreas Adipositas"
        foodArray!!.add(i3)
        val i4 = FoodItem("Salz")
        i4.type = 0
        i4.manufacturer = "Bad Reichenhaller"
        i4.setCategory(3, "Gewürz")
        i4.calories = 1f
        i4.authorName = "Andreas Adipositas"
        foodArray!!.add(i4)
    }

    companion object {

        fun newInstance(): FoodItemFragment {
            val fragment = FoodItemFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
